components/net/http_cache.rs

```diff
@@ -0,0 +1,488 @@
+                // Other status codes can only use heuristic freshness if the public cache directive is present.
+                if let Some(&header::CacheControl(ref directives)) = response.headers.get::<header::CacheControl>() {
+                    let has_public_directive = directives.iter().any(|directive| {
+                        header::CacheDirective::Public == *directive
+                    });
+                    if has_public_directive {
+                        return heuristic_freshness;
+                    }
+                }
+            },
+        }
+    }
+    // Requires validation upon first use as default.
+    Duration::seconds(0i64)
+}
+
+/// Request Cache-Control Directives https://tools.ietf.org/html/rfc7234#section-5.2.1
+fn get_expiry_adjustment_from_request_headers(request: &Request, expires: Duration) -> Duration {
+    if let Some(directive_data) = request.headers.get_raw("cache-control") {
+        let directives_string = String::from_utf8(directive_data[0].to_vec()).unwrap();
+        let directives: Vec<&str> = directives_string.split(",").collect();
+        for directive in directives {
+            let directive_info: Vec<&str> = directive.split("=").collect();
+            match directive_info[0] {
+                "max-stale" => {
+                    let seconds = String::from_str(directive_info[1]).unwrap();
+                    return expires + Duration::seconds(seconds.parse::<i64>().unwrap());
+                },
+                "max-age" => {
+                    let seconds = String::from_str(directive_info[1]).unwrap();
+                    let max_age = Duration::seconds(seconds.parse::<i64>().unwrap());
+                    if expires > max_age {
+                        return Duration::min_value();
+                    }
+                    return expires - max_age;
+                },
+                "min-fresh" => {
+                    let seconds = String::from_str(directive_info[1]).unwrap();
+                    let min_fresh = Duration::seconds(seconds.parse::<i64>().unwrap());
+                    if expires < min_fresh {
+                        return Duration::min_value();
+                    }
+                    return expires - min_fresh;
+                },
+                "no-cache" | "no-store" => return Duration::min_value(),
+                _ => {}
+            }
+        }
+    }
+    expires
+}
+
+/// Create a CachedResponse from a request and a CachedResource.
+fn create_cached_response(request: &Request, cached_resource: &CachedResource)
+    -> CachedResponse {
+    let mut response = Response::new(cached_resource.metadata.final_url.clone());
+    response.headers = cached_resource.metadata.headers.lock().unwrap().clone();
+    response.body = cached_resource.body.clone();
+    response.status = cached_resource.status.clone();
+    response.raw_status = cached_resource.raw_status.clone();
+    response.url_list = cached_resource.url_list.clone();
```

Comment:
We'll want to copy:
* the https state of the cached response
* the referrer and referrer policy of the request into the response

Reply:
OK

Reply:
done, see lines below.